src/javaharness/java/arcs/android/demo/service/AutofillRenderer.java

```diff
@@ -0,0 +1,84 @@
+package arcs.android.demo.service;
+
+import android.view.autofill.AutofillId;
+import android.service.autofill.Dataset;
+import android.service.autofill.FillCallback;
+import android.service.autofill.FillResponse;
+import android.view.autofill.AutofillValue;
+import android.widget.RemoteViews;
+import arcs.api.UiRenderer;
+import arcs.api.PortableJson;
+
+import java.util.HashMap;
+import java.util.Map;
+import javax.inject.Inject;
+
+public class AutofillRenderer implements UiRenderer {
+
+  public static class SlotInfo {
+    final String packageName;
+    final AutofillId autofillId;
+    final FillCallback fillCallback;
+    final Runnable customCallback;
+    SlotInfo(String packageName, AutofillId autofillId, FillCallback fillCallback, Runnable customCallback) {
+      this.packageName = packageName;
+      this.autofillId = autofillId;
+      this.fillCallback = fillCallback;
+      this.customCallback = customCallback;
+    }
+  }
+  private final Map<String, SlotInfo> slotById = new HashMap<>();
+
+  @Inject
+  AutofillRenderer() {
+  }
+
+  @Override
+  public boolean render(PortableJson content) {
+    PortableJson data = content.getObject("data");
+    String slotId = data.getString("outputSlotId");
```

Comment:
There's a lot of machinery here for pairing up slotIds with their callbacks. I suspect we could probably pull a lot of this out so that clients can write much simpler renderers, e.g. provide a new renderer for a given (modality, slot)-pair. Maybe the UiBroker can let you register renderers for particular slots as well as more general renderers for all slots?

Nothing to do now in this PR, just a thought.

Reply:
Agreed. 
It is already so much better after addressing your comments. And we need at least one more renderer to better understand what can be factored out.